content/blog/blog-2020-07-10-webhook-development.md

```diff
@@ -0,0 +1,153 @@
+Once ngrok is installed and registered (needed for 443), you should be able to run `ngrok http 443`
+
+This results in a running process that should forwarding details such as:
+
+```
+Web Interface                 http://127.0.0.1:4040                                                                                                                           
+Forwarding                    http://1a3430057023.ngrok.io -> https://localhost:443                                                                                           
+Forwarding                    https://1a3430057023.ngrok.io -> https://localhost:443  
+```
+It also has web socket at 4040 as indicated.  Which means we can get the tunnel urls in an automated fashion via: `curl -s localhost:4040/api/tunnels`.  The order of return of the end points changes.  In order to reliably get the https url, it requires some jq magic.  `curl -s localhost:4040/api/tunnels | jq '.tunnels[] | select(.proto == "https") | .public_url' -r`
+
+### Working with MutatingWebhookConfiguration
+
+The default configuration for KUDO looks like this (abbreviated by removing "caBundle"):
+
+```
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: MutatingWebhookConfiguration
+metadata:
+  creationTimestamp: null
+  name: kudo-manager-instance-admission-webhook-config
+webhooks:
+- clientConfig:
+    caBundle: abbreviated
+    service:
+      name: kudo-controller-manager-service
+      namespace: kudo-system
+      path: /admit-kudo-dev-v1beta1-instance
+  failurePolicy: Fail
+  matchPolicy: Equivalent
+  name: instance-admission.kudo.dev
+  rules:
+  - apiGroups:
+    - kudo.dev
+    apiVersions:
+    - v1beta1
+    operations:
+    - CREATE
+    - UPDATE
+    resources:
+    - instances
+    scope: Namespaced
+  sideEffects: None
+```
+
+For a webhook, there are 2 ways to configure the callback.  The default for KUDO and is likely the most common in-cluster configuration is to configure the `clientConfig.service` configuration.  The alternative option is via the `clientConfig.url`.  It is NOT valid to have both of these configured at the same time.  Which also means it is challenging to patch update an existing configuration which currently has `.service` with a `.url`.  It is best to delete the configuration in this case: `kubectl delete MutatingWebhookConfiguration kudo-manager-instance-admission-webhook-config`.
+
+We need to switch from the `clientConfig.service` approach above to the `clientConfig.url` approach using the current ngrok configuration by replacing:
```

Comment:
I would probably break up the alternative option into its own section – I wasn't really sure at first glance whether I *need to* replace this section, or whether I *can*, in case I want to.